taskvine/src/worker/vine_cache.c

```diff
@@ -376,99 +387,163 @@ static int do_transfer( struct vine_cache *c, const char *source_url, const char
+				// pid is still going
+			}
+			else if(result<0){
+				debug(D_VINE, "wait4 on pid %d returned an error: %s",(int)f->pid,strerror(errno));	
+			}
+			else if(result>0){
+				if(f->type==VINE_CACHE_MINI_TASK){
+					vine_sandbox_stageout(f->process, c, manager);
+					f->process->task = 0;
+					vine_process_delete(f->process);
+				}
+				if(!WIFEXITED(status)){
+					exit_code = WTERMSIG(status);
+					debug(D_VINE, "transfer process (pid %d) exited abnormally with signal %d",f->pid, exit_code);
+					f->status = VINE_FILE_STATUS_FAILED;
+				} else {
+					exit_code = WEXITSTATUS(status);
+					debug(D_VINE, "transfer process for %s (pid %d) exited normally with exit code %d", cachename, f->pid, exit_code );
+					if(exit_code==1){	
+						debug(D_VINE, "transfer process for %s completed", cachename);
+						f->status = VINE_FILE_STATUS_READY;
+					} else {
+						debug(D_VINE, "transfer process for %s failed", cachename);
+						f->status = VINE_FILE_STATUS_FAILED;
+					}
+				}
+				if(f->status==VINE_FILE_STATUS_READY){
+					int64_t nbytes, nfiles;
+					char *cache_path = vine_cache_full_path(c,cachename);
+					f->stop_time = timestamp_get();
+					timestamp_t transfer_time = f->stop_time - f->start_time;
+					chmod(cache_path,f->mode);
+					if(path_disk_size_info_get(cache_path,&nbytes,&nfiles)==0) {
+						f->actual_size = nbytes;
+						debug(D_VINE,"cache: created %s with size %lld in %lld usec",cachename,(long long)f->actual_size,(long long)transfer_time);
+						vine_worker_send_cache_update(manager,cachename,f->actual_size,transfer_time,f->start_time);
+						f->status = VINE_FILE_STATUS_READY;
+					} else {
+						debug(D_VINE,"cache: command succeeded but did not create %s",cachename);
+						f->status = VINE_FILE_STATUS_FAILED;
+					}
+
+				} else {
+					debug(D_VINE,"cache: unable to create %s",cachename);
+				}
+			}
+		}	
+	}
+	return 1;
+}
 
 /*
 Ensure that a given cached entry is fully materialized in the cache,
 downloading files or executing commands as needed.  If complete, return
-true, otherwise return false.
+VINE_FILE_STATUS_READY, If downloading return VINE_FILE_STATUS_PROCESSING.
+On failure return VINE_FILE_STATUS_FAILED.
 
-It is a little odd that the manager link is passed as an argument here,
-but it is needed in order to send back the necessary update/invalid messages.
 */
 
-int vine_cache_ensure( struct vine_cache *c, const char *cachename, struct link *manager )
+vine_file_status_type_t vine_cache_ensure( struct vine_cache *c, const char *cachename)
 {
-	if(!strcmp(cachename,"0")) return 1;
+	if(!strcmp(cachename,"0")) return VINE_FILE_STATUS_READY;
 
 	struct cache_file *f = hash_table_lookup(c->table,cachename);
 	if(!f) {
 		debug(D_VINE,"cache: %s is unknown, perhaps it failed to transfer earlier?",cachename);
-		return 0;
+		return VINE_FILE_STATUS_FAILED;
 	}
 
-	if(f->complete) {
-		/* File is already present in the cache. */
-		return 1;
+	/* File is already present in the cache. */
```

Comment:
As a general rule, when you are examining the states of an enumeration, use a `switch` and address all of the cases.
This is a defensive measure that will make it easier to maintain the code if/when the enumeration adds or removes cases.